swarm/storage/resource.go

```diff
@@ -0,0 +1,528 @@
+// successfully retrieved version is copied to the corresponding resources map entry
+// and returned.
+//
+// If refresh is set to true, the resource data will be reloaded from the resource update
+// root chunk.
+// It is the callers responsibility to make sure that this chunk exists (if the resource
+// update root data was retrieved externally, it typically doesn't)
+func (self *ResourceHandler) OpenResource(name string, refresh bool) (*resource, error) {
+
+	rsrc := &resource{}
+
+	// if the resource is not known to this session we must load it
+	// if refresh is set, we force load
+	if _, ok := self.resources[name]; !ok || refresh {
+
+		// make sure our ens identifier is idna safe
+		validname, err := idna.ToASCII(name)
+		if err != nil {
+			return nil, err
+		}
+		rsrc.name = validname
+		rsrc.ensname = ens.EnsNode(validname)
+
+		// get the root info chunk and update the cached value
+		chunk, err := self.Get(Key(rsrc.ensname[:]))
+		if err != nil {
+			return nil, err
+		}
+
+		// sanity check for chunk data
+		// data is prefixed by 8 bytes of size
+		if len(chunk.SData) < 24 {
+			return nil, fmt.Errorf("Invalid chunk length %d", len(chunk.SData))
+		} else {
+			chunklength := binary.LittleEndian.Uint64(chunk.SData[:8])
+			if chunklength != uint64(16) {
+				return nil, fmt.Errorf("Invalid chunk length header %d", chunklength)
+			}
+		}
+		rsrc.startblock = binary.LittleEndian.Uint64(chunk.SData[8:16])
+		rsrc.frequency = binary.LittleEndian.Uint64(chunk.SData[16:])
+	} else {
+		rsrc.name = self.resources[name].name
+		rsrc.ensname = self.resources[name].ensname
+		rsrc.startblock = self.resources[name].startblock
+		rsrc.frequency = self.resources[name].frequency
+	}
+
+	// get our blockheight at this time and the next block of the update period
+	currentblock, err := self.getBlock()
+	if err != nil {
+		return nil, err
+	}
+	nextblock := getNextBlock(rsrc.startblock, currentblock, rsrc.frequency)
+
+	// start from the last possible block period, and iterate previous ones until we find a match
+	// if we hit startblock we're out of options
+	version := uint64(1)
+	for nextblock > rsrc.startblock {
+		key := self.resourceHash(rsrc.ensname, nextblock, version)
+		chunk, err := self.Get(key)
+		if err == nil {
+			// check if we have versions > 1. If a version fails, the previous version is used and returned.
+			log.Trace("rsrc update version 1 found, checking for version updates", "nextblock", nextblock, "key", key)
+			for {
+				newversion := version + 1
+				key := self.resourceHash(rsrc.ensname, nextblock, newversion)
+				newchunk, err := self.Get(key)
+				if err != nil {
+					// rsrc update data chunks are total hacks
+					// and have no size prefix :D
+					err := self.verifyContent(chunk.SData)
+					if err != nil {
+						return nil, err
+					}
+					// update our rsrcs entry map
+					rsrc.lastblock = nextblock
+					rsrc.version = version
+					rsrc.data = make([]byte, len(chunk.SData)-SIGNATURE_LENGTH)
```

Comment:
We should probably check for `len(chunk.SData) >= 24` again, since we performed a new `Get` operation.

Reply:
it's not the same kind of chunk :) the length is checked in `getContentAccount`